### ISSUE TITLE
Preserve order and prioritize test cases in webdriver.io

### ISSUE THREAD

**Issue by Arm91** [webdriverio/webdriverio#4845]
**Is your feature request related to a problem? Please describe.**
As the number of test cases grows it becomes necessary to preserve the order of running test files and prioritize test cases which to run first and which to run last.
There are some apps, that an action in one test case can affect actions on other test cases.

**Describe the solution you'd like**
It will be good to have a config file, where the user can preserve the order of running test files and also could give priorities to test cases inside test cases.

**Describe alternatives you've considered**
A very good example of preserving  test case ordering and prioritizing is implemented in TestNG

> By default, TestNG will run your tests in the order they are found in the XML file. If you want the classes and methods listed in this file to be run in an unpredictable order, set the preserve-order attribute to false

As for giving priorities to test cases please see [this](https://www.guru99.com/test-case-priority-testng.html) documentation.

 
Also there is a TestNG dependsOnMethod, which allows to run some test cases only after certain test cases are run and all conditions have passes, a good example is [here](https://www.journaldev.com/21389/testng-dependency-dependsonmethods-dependsongroups).

**Additional context**
The official [documentation](https://testng.org/doc/documentation-main.html) of TestNG

**Comment by erwinheitzman** [comment]
You can already assign an order by creating suites, using describe blocks or sorting the specs array. Would these solutions solve your issue?

**Comment by Arm91** [comment]
@erwinheitzman , do you mean if I specify every file in a suite in WDIO, it will run in that order?
example:
```
suites: {
        fullRegression: [
            './test/specs/file10.spec.js',
            './test/otherFolder/file4.spec.js',
           './test/specs/file1.spec.js',
           './test/anotherFolder/file7.spec.js'
        ],
```
First will run 10, then 4, then 1 and then 7? And all of them are from different folders? (but all in main ./test folder)


Also a question regarding hooks, will this work in the specified order?

```
beforeEach(()=>{
})
Describe('Test Suite',()=>{
    before(()=>{
    })
    it('Test Case 1 ', ()=>{
    })
    it('Test Case 2 ', ()=>{
    })
    after(()=>{
    })
})
afterEach(()=>{
})

```

> sorting the specs array.

what is this one? Any documentation?

**Comment by erwinheitzman** [comment]
Regarding the suites, I believe it works like this but I cannot tell for sure right now.

Regarding the hooks, yes that will work. However, keep the following in mind:
```
describe('1d', () => {

  beforeEach(() => {
    console.log('first');
  });

  describe('blaat', () => {
    beforeEach(() => {
      console.log('second');
    });

    describe('1d', () => {
      beforeEach(() => {
        console.log('third');
      });
  
      it('1i', () => {
          console.log('fourth')
      });
    });
  
    describe('2d', () => {
      beforeEach(() => {
        console.log('fifth');
      });
  
      it('2i', () => {
        console.log('sixth');
      });
    });
  });
});
```

This will output:
```
[0-0] first
[0-0] second
[0-0] third
[0-0] fourth
[0-0] first
[0-0] second
fifth
sixth
```

Regarding the specs array. In your config file you have the specs property. You could generate an array of spec files (test files) and sort them in whatever order you'd like like so:
```
// wdio.conf.js
function getSpecFiles() {
  // logic that retrieves the filenames of your testfiles (like recursively using readFileSync)
}

exports.config = {
  // ...
  specs: getSpecFiles().sort(a =>{
    // logic to sort
  })
  // ...
}
```

**Comment by erwinheitzman** [comment]
@Arm91 can you give an update if my comment helped you in resolving your issue?

**Comment by Arm91** [comment]
@erwinheitzman ,

I can officially confirm that test files are run as specified in test suite
```
suites: {
        fullRegression: [
            './test/specs/file10.spec.js',
            './test/otherFolder/file4.spec.js',
           './test/specs/file1.spec.js',
           './test/anotherFolder/file7.spec.js'
        ],
```
This will run 10, then 4, then 1 and then 7 as long as the
`maxInstances: 1,` is 1. 
If it is more, then it depends on the length of the test case steps.
For example maxInstances is 2: 10 and 4 will run together, but if 4 finishes first, 1 will run while 10 is still in progress. If maxInstances is more, then everything becomes more complicated and pretty much uncontrollable, because the speed of test cases varies.

So this is not a very good workaround, but at least something for now.

As for hooks, I played with it and now just login once and logout once in a file to make them faster
```
describe('Global Suite', () => {
  before(() => {
    console.log('>>>> inside before');
  });
  describe('Test Suite 1', () => {
    beforeEach(() => {
      console.log('>>>> Inside beforeEach 1');
    });
    it('Test Case 1.1', () => {
      console.log('>>>> Inside Test Case 1.1');
    });
    it('Test Case 1.2', () => {
      console.log('>>>> Inside Test Case 1.2');
    });
    afterEach(() => {
      console.log('>>>> Inside afterEach 1');
    });
  });
  describe('Test Suite 2', () => {
    beforeEach(() => {
      console.log('>>>> Inside beforeEach 2');
    });
    it('Test Case 2.1', () => {
      console.log('>>>> Inside Test Case 2.1');
    });
    it('Test Case 2.2', () => {
      console.log('>>>> Inside Test Case 2.2');
    });
    afterEach(() => {
      console.log('>>>> Inside afterEach 2');
    });
  });
  after(() => {
    console.log('>>>> Inside after ');
  });
});
```

The output for this is
```
[0-0] >>>> inside before
[0-0] >>>> Inside beforeEach 1
[0-0] >>>> Inside Test Case 1.1
[0-0] >>>> Inside afterEach 1
[0-0] >>>> Inside beforeEach 1
[0-0] >>>> Inside Test Case 1.2
[0-0] >>>> Inside afterEach 1
[0-0] >>>> Inside beforeEach 2
[0-0] >>>> Inside Test Case 2.1
[0-0] >>>> Inside afterEach 2
[0-0] >>>> Inside beforeEach 2
[0-0] >>>> Inside Test Case 2.2
[0-0] >>>> Inside afterEach 2
[0-0] >>>> Inside after 
```


And finally as for:
> generate an array of spec files (test files) and sort them in whatever order you'd like

I have too many folders and too many files with different naming, sorting them is pretty much impossible, so this is not a solution for me.

**Comment by CrispusDH** [comment]
you can split running on several scripts. For example:
```
  "scripts": {
    "first": "run my first suite or tests...",
    "second": "run my second suite or tests...a",
     .....
    "test": "npm run first && npm run second....",
}
```

**Comment by christian-bromann** [comment]
@Arm91 it seems that glob preserves the order of files it finds, so it you have `{n}_mocha.test.js` created and defined in your config as follows:

```js
specs: [path.resolve(__dirname, '*_mocha.test.js')],
```

It will always run the spec files in right order: 1_mocha.test.js, 2_mocha.test.js and 3_mocha.test.js. If you run tests in parallel you will of course experience that files are run in multiples and the order is not file by file. Framework handle this differently , e.g. the Jasmine framework has a [`random`](https://www.npmjs.com/package/@wdio/jasmine-framework#random) option that allows to run test in serial or randomly.

I will close the issue as I see that spec execution order is being preserved in the wdio testrunner. I am interested what you have in mind on prioritization of test files. I know that other testrunner have such functionalities but this would be good to propose such an idea in a new issue. Thanks!
